Fix StyleSheet 'textAlign' for AndroidTextInput. Closes #2702

trave7er opened the pull request:
change `setTextAlign` and `setTextAlignVertical` to receive argument of type `String` (the same as in `StyleSheet`), so that native props and stylesheet props are calling the same @ReactMethod

- add demo (may not be necessary)

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -59,6 +61,24 @@
   private static final InputFilter[] EMPTY_FILTERS = new InputFilter[0];
 
+  private static final Map<String, Map<String, Integer>> TEXT_ALIGN_CONSTANTS;
+  static {
```
you don't need a static block for this, you can just continue with this assignment on the line above

kmagiera commented on Libraries/Components/TextInput/TextInput.js:
```diff
@@ -517,8 +513,8 @@ var TextInput = React.createClass({
         autoCorrect={this.props.autoCorrect}
-        textAlign={textAlign}
-        textAlignVertical={textAlignVertical}
+        textAlign={this.props.textAlign}
```
I think I'd still prefer this to be moved into styles as opposed to being a view prop. Just for the sake of consistency with the Text element.

Note that currently `style` property of `TextInput` element use the same `propType` as `Text` (see [here](https://github.com/facebook/react-native/blob/c11dce36e6488c77eb16c121ea7cfe497c26ff4a/Libraries/Components/TextInput/TextInput.js#L314)), which allows `textAlign` (see [here](https://github.com/facebook/react-native/blob/62e8ddc20561a39c3c839ab9f83c95493df117c0/Libraries/Text/TextStylePropTypes.js#L40)). This means that using `textAlign` for `TextInput` in `style` won't result in a warning.

I think we should add `textAlignVertical` to the `style` definition (with an annotation it's android-only) and remove it from `propTypes` definition for `TextInput`

kmagiera commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -189,13 +209,17 @@ public void setUnderlineColor(ReactEditText view, @Nullable Integer underlineCol
   @ReactProp(name = "textAlignVertical")
-  public void setTextAlignVertical(ReactEditText view, int gravity) {
+  public void setTextAlignVertical(ReactEditText view, @Nullable String textAlignVertical) {
+    Map<String, Integer> constants = TEXT_ALIGN_CONSTANTS.get("textAlignVertical");
```
I wouldn't use hash maps for this in java. They tend not to be very efficient (compering to alternative approaches) when you have a small number of keys. I'd recommend to at least have a separate maps for `TEXT_ALIGN_CONSTANTS` and `TEXT_ALIGN_VERTICAL_CONSTANTS` or even resign from the secondary level map here and just having a bunch of if blocks similarily to how it's handled here: [ReactTextViewManager.java](https://github.com/facebook/react-native/blob/a0268a7bfc8000b5297d2b50f81e000d1f479c76/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextViewManager.java#L60)

mkonicek commented:
Thanks for the review @kmagiera!

trave7er commented:
@kmagiera , I updated the diff based on your comments. Let me know what you think!

Some highlights of updates:

 - removed the use of `HashMap`
 - removed Android native props for `textAlign` & `textAlignVertical`
 - added `textAlignVertical` to `TextStyleProps`
 - added `TEXT_ALIGN_VERTICAL` constant to `ViewProps`

kmagiera commented:
This looks very good. Thanks @traver7er!

I believe this should be marked as a breaking change and someone should update callsites in Fb internal apps. What needs to be done is just to move textAlign and textVerticalAlign out of view props and to the styles object cc @mkonocek

mkonicek commented:
Adding to my TODO-list of PRs.

trave7er commented:
was going to also update the docs, and found it's already updated. Thanks!

Let me know if you need me to update anything,

trave7er commented:
@mkonicek @kmagiera  a quick note, I named the prop "textAlignVertical" instead of "textVerticalAlign", do you want me to change it?

satya164 commented:
@trave7er I think `textAlignVertical` is better.

GantMan commented:
:+1:  on this one.  I was able to put off the "Android Version" of an app to 2016, but I'd love :shipit: soon.

christopherdro commented:
@trave7er Can you resolve the merge conflicts so we can try and get this merged in?

trave7er commented:
@christopherdro updated

gr4yscale commented:
just ran into this issue myself - hoping this gets merged today!

jsierles commented:
@trave7er Could you also squash commits?

mkonicek commented:
Actually I think shipit squashes the commits, let's try it:

@facebook-github-bot shipit

mkonicek commented:
Yup, looks squashed, landing now :)

gr4yscale commented:
Any news on this?

mkonicek commented:
I need to fix an internal test, will cherry-pick this to 0.19.

mkonicek commented:
Merging again now. Please don't update the PR :)

mkonicek commented:
:rocket:
